domain/secretbackend/state/state.go

```diff
@@ -720,6 +795,279 @@ func (s *State) getModelCloudAndCredential(
+// or [modelerrors.NotFound] if the model does not exist,
+// or [secretbackenderrors.RefCountAlreadyExists] if the reference already exists.
+// If the ValueRef is nil, the internal controller backend is used.
+// It returns a rollback function which can be used to revert the changes.
+func (s *State) AddSecretBackendReference(
+	ctx context.Context, valueRef *secrets.ValueRef, modelID coremodel.UUID, revisionID uuid.UUID,
+) (func() error, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var backendID string
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		backendID, err = s.getSecretBackendID(ctx, tx, valueRef)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err := s.addSecretBackendReference(ctx, tx, backendID, modelID, revisionID)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return func() error {
+		err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			input := SecretBackendReference{
+				BackendID:        backendID,
+				SecretRevisionID: revisionID.String(),
+			}
+			stmt, err := s.Prepare(`
+DELETE FROM secret_backend_reference
+WHERE  secret_backend_uuid = $SecretBackendReference.secret_backend_uuid
+    AND secret_revision_uuid = $SecretBackendReference.secret_revision_uuid`, input)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = tx.Query(ctx, stmt, input).Run()
+			if errors.Is(err, sql.ErrNoRows) {
+				return nil
+			}
+			if err != nil {
+				return fmt.Errorf(
+					"cannot revert secret backend reference for secret backend %q secret revision %q: %w",
+					backendID, revisionID, err,
+				)
+			}
+			return nil
+		})
+		return errors.Trace(err)
+	}, nil
+}
+
+func (s *State) getSecretBackendID(ctx context.Context, tx *sqlair.TX, valueRef *secrets.ValueRef) (string, error) {
+	arg := secretbackend.BackendIdentifier{}
+	if valueRef != nil {
+		// We want to check if the backend exists for the given ID.
+		arg.ID = valueRef.BackendID
+	} else {
+		arg.Name = juju.BackendName
+	}
+
+	backend, err := s.getSecretBackend(ctx, tx, arg)
+	if err != nil {
+		return "", fmt.Errorf("cannot get secret backend ID for %q: %w", arg, err)
+	}
+	return backend.ID, nil
+}
+
+func (s *State) addSecretBackendReference(
+	ctx context.Context, tx *sqlair.TX, backendID string, modelID coremodel.UUID, revisionID uuid.UUID,
+) error {
+	input := SecretBackendReference{
+		BackendID:        backendID,
+		ModelID:          modelID,
+		SecretRevisionID: revisionID.String(),
+	}
+
+	stmt, err := s.Prepare(`
+INSERT INTO secret_backend_reference (*)
+VALUES ($SecretBackendReference.*)`, input)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = tx.Query(ctx, stmt, input).Run()
+	if database.IsErrConstraintForeignKey(err) {
+		return fmt.Errorf("%w: model %q", modelerrors.NotFound, input.ModelID)
+	}
+	if database.IsErrConstraintPrimaryKey(err) {
```

Comment:
This is never expected to happen right? As with above, should have a comment to explain why this inefficient mechanism is used.